### PR TITLE
`Programming exercises`: Store the used branch in the database

### PR DESCRIPTION
**Pull request by JohannesStoehr**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Changes affecting Programming Exercises
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

### Motivation and Context

Retrieving the default branch from the VCS for many git operations is not performant and can slow down the system.

### Description

This PR stores the default Branch for the used repositories in the database of Artemis and only retrieves them from the VCS if necessary.
This PR contains a database change. Make surre to delete all newly created programming exercises and participations after testing!

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Programming Exercise

1. Log in to Artemis
2. Create programming exercise
3. Make sure that everything is still working in the programming exercise:
- Online Editor for Students
- - Online Editor for Instructors
- Cloning to the local IDE, pushing and. receiving feedback
- Importing (Especially If you still have an old exercise with "master" as default branch)
- ...

#### Code Review
- [ ] Review 1
- [ ] Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2

### Test Coverage

| Class/File | Branch | Line |
|------------|-------:|-----:|

### CODE REVIEW

**Comment by dfuchss** [comment]
@JohannesStoehr do you think we could try to add that to the next mile stone? This feature would help us a lot :)

**Review by nairaarnez — COMMENTED** [review]
While trying to use the online editor I couldn't submit the exercise. It Builded and tested for around 15 minutes now when I pressed the submit button.
image

**Review by Dominik-Weinzierl — COMMENTED** [review]
The online editor works on TS2 with this PR - could you please test it again @nairaarnez :thinking:

**Review by willergeorg — APPROVED** [review]
Tested on TS2. Working with online editor worked fine. I could not reproduce the issues reported by @nairaarnez.

**Comment by JohannesStoehr** [comment]
> While trying to use the online editor I couldn't submit the exercise. It Builded and tested for around 15 minutes now when I pressed the submit button.
> image

There were some problems with the Buildagents on TS1. This is probably related.

**Review by cremertim — DISMISSED** [review]
Tested on ts2, worked as expected. Was not able to reproduce the bug either

**Review by nairaarnez — COMMENTED** [review]
I just deployed another PR and I become the same message, it is not because of your PR Johannes! Does someone happend to know why do I get this message?

File: src/main/java/de/tum/in/www1/artemis/service/connectors/VersionControlService.java [hunk 1]
```diff
@@ -186,4 +188,30 @@ default String getDefaultBranchOfRepository(String projectKey, String repository
      * @return The health of the VCS service containing if it is up and running and any additional data, or the throwing exception otherwise
      */
     ConnectorHealth health();
+
+    /**
+     * Get the default branch used in the participation or retrieves it from the VCS if not present in the database
+     * @param participation The participation to get the default branch from
+     * @return The default branch used by this participation
+     */
+    String getOrRetrieveDefaultBranchOfParticipation(ProgrammingExerciseParticipation participation);
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/VersionControlService.java:197]: Can we rename this to `getOrRetrieveBranchOfParticipation`
(same for the methods below)

**Comment by dfuchss** [comment]
I've created a project with old "master" branch. 
Afterwards, I've stopped artemis and delted the repos on the local clone of artemis.

Afterwards, I've changed the default branch to "main" and updated to this PR.

Now I get internal server errors (gitlab/jenkins setup) while opening the "Edit in Editor" view as instructor:

```
2022-05-20 09:12:11.114 ERROR 1 --- [nio-8080-exec-9] o.z.problem.spring.common.AdviceTraits   : Internal Server Error

java.lang.IllegalArgumentException: null
        at de.tum.in.www1.artemis.web.rest.repository.RepositoryProgrammingExerciseParticipationResource.getOrRetrieveDefaultBranchOfDomainObject(RepositoryProgrammingExerciseParticipationResource.java:165)
        at de.tum.in.www1.artemis.web.rest.repository.RepositoryResource.getStatus(RepositoryResource.java:294)
        at de.tum.in.www1.artemis.web.rest.repository.RepositoryProgrammingExerciseParticipationResource.getStatus(RepositoryProgrammingExerciseParticipationResource.java:348)
        at de.tum.in.www1.artemis.web.rest.repository.RepositoryProgrammingExerciseParticipationResource$$FastClassBySpringCGLIB$$9ebf5bf.invoke(<generated>)
        at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
```

**Review by dfuchss — CHANGES_REQUESTED** [review]
See comment

**Review by dfuchss — APPROVED** [review]
Code works on our Test System

**Pull request merged**